### PR TITLE
Major refactoring of mpconfig and makefiles; compile only what's used; easy on/off of modules

### PR DESCRIPTION
**Pull request by dhalbert**
This is a large refactoring of `mpconfig*.{h,mk}` settings to share as much as possible and to make enabling and disabling builtin modules easier.

There's a new directory `ports/circuitpy-common/` containing settings that can be shared across ports:
  - `mpconfig_circuitpy.h`: refactoring of common aspects of `{atmel-samd,nrf}/mpconfigport.h`
  - `mpconfig_circuitpy.mk`: refactoring of common aspects of `{atmel-samd,nrf}/mpconfigport.mk`
  - `circuitpy_defns.mk`: refactoring of common elements of `{atmel-samd,nrf}/Makefile`
  -  (there was a common Makefile rules file, but the builds for atmel-samd and nrf were a little too different to merge for now)

There are new makefile variables `CIRCUITPY_AUDIOIO`, `CIRCUITPY_DISPLAYIO`, `CIRCUITPY_OS`, etc. with values 0 or 1. There's one for nearly every builtin module in `shared-bindings`, though some dependencies are handled automatically (e.g. `displayio` implies `terminalio`). So each and every builtin module can be turned on or off as needed. There are sensible defaults. There are `CIRCUITPY_SMALL_BUILD` and `CIRCUITPY_FULL_BUILD` variables which imply sets of settings. `CIRCUITPY_SMALL_BUILD` is for most M0 non-Express boards, and is set to `1` in the `mpconfigboard.mk` files for those boards. `CIRCUITPY_FULL_BUILD` is the default.

The makefile variables above also set corresponding `#define` macros, set by `-DCIRCUITPY_xxx=1` or `0`, so you can do `#if CIRCUITPY_AUDIO`, etc.

This allows removing all the _ad hoc_ stuff we had before like `PIRKEY_M0` and `EXCLUDE_PIXELBUF` macros. It's a lot cleaner to set up a special build with only certain features.

Based on the `CIRCUITPY_xxx` values, only the source files needed for the functionality are compiled, and unused ones are completely absent. See the use of `$(filter ...)` in `mpconfig_circuitpy.mk`. This means that there's no dead code: even `-flto` was including some unused code. So this saves significant space for small builds. It also makes small builds compile faster (~20% faster, maybe). For large builds, most features are on anyway.

The `nrf/Makefile` has been revised to be much closer to the `atmel-samd/Makefile`, with most of the same warnings, compiler settings, etc. I haven't yet tried `-flto` on `nrf`, but I will later. That was broken previously, and that problem might still be present. (It might be due to how the ISR vector is defined.) A couple of gcc warnings had to be turned off due to compilation issues with `nrfx`.

Due to the increased set of warnings in the `nrf` port, I had to make a number of small code revisions that fix problems caught by the new warnings.

A few more things that could still use cleanup or consideration:

`CIRCUITPY_NETWORK` still has some _ad hoc_ settings checks about which Ethernet adapter to use. This could probably be made a little cleaner.

With the increased space available on small builds, we could consider turning on a few more features to make the small and regular builds closer in functionality, such as `MICROPY_CPYTHON_COMPAT`, a few missing string functions, etc.

I Need to check about `MICROPY_MODULE_WEAK_LINKS`, which I think is no longer necessary but may need a bit of cleanup if it's still used.

Testing:

Built all the significant `atmel-samd` boards and tried a bunch on real boards. Travis should be relatively happy.

Built Feather nRF52840 and tested simple BLE functionality.

### CODE REVIEW

File: ports/atmel-samd/mpconfigport.h [hunk 1]
```diff
@@ -181,302 +64,31 @@ typedef long mp_off_t;
 #ifdef SAMD51
 #define CIRCUITPY_MCU_FAMILY                        samd51
 #define MICROPY_PY_SYS_PLATFORM                     "MicroChip SAMD51"
-#define PORT_HEAP_SIZE                              (0x20000) // 128KiB
+// 128KiB heap
+#define PORT_HEAP_SIZE                              (0x20000)
```
Comment (tannewt) [inline ports/atmel-samd/mpconfigport.h]: I don't think we use this at all. It should all be dynamic now.

**Comment by dhalbert** [comment]
>I'm not a fan of `ports/circuitpy-common` though. I'd rather have those three files in `py` where the micropython equivalents are. They could just start with `circuitpy` to separate them out.

That's fine -- I wasn't completely happy with this, and was going to ask you about it but forgot.

File: ports/atmel-samd/mpconfigport.h (hunk 1 above)
Comment (dhalbert) [reply ports/atmel-samd/mpconfigport.h]: Yep, defined a lot of places but not used at all. I'll remove

**Comment by sommersoft** [comment]
I dig this! Much cleaner, and more modular. Thanks @dhalbert!!

**Comment by notro** [comment]
> I Need to check about `MICROPY_MODULE_WEAK_LINKS`, which I think is no longer necessary but may need a bit of cleanup if it's still used.

The CPython std library port I'm working on requires this. It makes it possible to extend bare bones builtin modules using pure python.
Most noteable are the [os](https://github.com/notro/tmp_CircuitPython_stdlib/blob/master/lib/os.py) and [time](https://github.com/notro/tmp_CircuitPython_stdlib/blob/master/lib/time.py) modules that I have ported.

Another word for WEAK_LINKS is module alias.

**Comment by dhalbert** [comment]
@notro I'll open an issue about this. Yes, WEAK_LINKS is not a great phrase, since it overlaps with linker weak links, and that was confusing for a bit.

**Pull request merged**